### PR TITLE
Gene conversion breakpoint insertion and verification against ms

### PR DESCRIPTION
**Pull request by fbaumdicker**
This contains the breakpoint insertion to make the comparison to ms possible and alters the ms code to record recombination and gene conversion events separately.

### CODE REVIEW

File: lib/msprime.c [hunk 1]
```diff
@@ -238,11 +238,13 @@ msp_set_gene_conversion_rate(msp_t *self, double rate, double track_length)
          * always maps to the same genetic distance */
 
         if (track_length < 0 || track_length > self->recomb_map->sequence_length) {
+            printf("tl > recomb_map seq length");
```
Comment (jeromekelleher) [inline lib/msprime.c]: Strayt debug printf here

File: lib/msprime.c [hunk 2]
```diff
@@ -238,11 +238,13 @@ msp_set_gene_conversion_rate(msp_t *self, double rate, double track_length)
...
             ret = MSP_ERR_BAD_PARAM_VALUE;
             goto out;
         }
         genetic_track_length = recomb_map_phys_to_genetic(self->recomb_map, track_length);
         if (genetic_track_length < 1.0) {
+            printf("genetic track length < 1");
```
Comment (jeromekelleher) [inline lib/msprime.c]: Same here

File: lib/msprime.c [hunk 3]
```diff
@@ -1504,6 +1507,14 @@ msp_cut_right_break(msp_t *self, segment_t *lhs_tail, segment_t *y, segment_t *n
     y->next = NULL;
     y->right = track_end;
     fenwick_increment(&self->links[label], y->id, (int64_t) track_end - (int64_t) new_segment->right);
+    if (msp_has_breakpoint(self, (uint32_t) track_end)) {
```
Comment (jeromekelleher) [inline lib/msprime.c]: Remove redundant else here since we're not using the first clause.

File: lib/msprime.c [hunk 4]
```diff
@@ -1504,6 +1507,14 @@ msp_cut_right_break(msp_t *self, segment_t *lhs_tail, segment_t *y, segment_t *n
...
+//             self->num_multiple_gc_breakpoints++;
+    } else {
+        ret = msp_insert_breakpoint(self, (uint32_t) track_end);
+        if (ret != 0) {
+            printf("Warning: Bad breakpoint insertion\n");
```
Comment (jeromekelleher) [inline lib/msprime.c]: This is an error, goto out

File: lib/msprime.c [hunk 5]
```diff
@@ -1565,6 +1577,23 @@ msp_gene_conversion_within_event(msp_t *self, label_id_t label)
             y->right = (uint32_t) k;
             fenwick_set_value(&self->links[label], z2->id, z2->right - y->right);
             fenwick_increment(&self->links[label], y->id, k - z2->right);
+            if (msp_has_breakpoint(self, (uint32_t) k)) {
+//             self->num_multiple_gc_breakpoints++;
```
Comment (jeromekelleher) [inline lib/msprime.c]: ditto redundant else here

File: lib/msprime.c [hunk 6]
```diff
@@ -1565,6 +1577,23 @@ msp_gene_conversion_within_event(msp_t *self, label_id_t label)
...
+            } else {
+                ret = msp_insert_breakpoint(self, (uint32_t) k);
+                if (ret != 0) {
+                    printf("Warning: Bad breakpoint insertion\n");
+                }
+            }
+
+            if (msp_has_breakpoint(self, (uint32_t) (k + tl) )) {
```
Comment (jeromekelleher) [inline lib/msprime.c]: Here also and elsewhere

File: lib/msprime.c [hunk 7]
```diff
@@ -1565,6 +1577,23 @@ msp_gene_conversion_within_event(msp_t *self, label_id_t label)
...
+//             self->num_multiple_gc_breakpoints++;
+            } else {
+                ret = msp_insert_breakpoint(self, (uint32_t) (k + tl) );
+                if (ret != 0) {
+                    printf("Warning: Bad breakpoint insertion\n");
```
Comment (jeromekelleher) [inline lib/msprime.c]: error, goto out

File: msprime/cli.py [hunk 8]
```diff
@@ -187,34 +187,25 @@ def print_trees(self, tree_sequence, output):
...
         simulation and write out a tree for each one.
         """
-        # breakpoints = self._simulator.breakpoints + [self._num_loci]
+        breakpoints = self._simulator.breakpoints + [self._num_loci]
         if self._num_loci == 1:
             tree = next(tree_sequence.trees())
             newick = tree.newick(precision=self._precision)
             print(newick, file=output)
         else:
-            # j = 0
```
Comment (jeromekelleher) [inline msprime/cli.py:196]: This needs cleaning up.

**Review by jeromekelleher — COMMENTED** [review]
Fantastic, thanks @fbaumdicker! Some comments above to clean things up a bit. Just needs a rebase then (probably best to squash the commits down too) and we're good to merge.

File: msprime/cli.py (hunk 8 above)
Comment (fbaumdicker) [reply msprime/cli.py:196]: I don't see the issue here?

Comment (jeromekelleher) [reply msprime/cli.py:196]: We want to get rid of the commented out code in this block and restore it to it's original state.

File: lib/msprime.c (hunk 1 above)
Comment (fbaumdicker) [reply lib/msprime.c]: removed

File: lib/msprime.c (hunk 2 above)
Comment (fbaumdicker) [reply lib/msprime.c]: removed

File: lib/msprime.c (hunk 3 above)
Comment (fbaumdicker) [reply lib/msprime.c]: done

File: lib/msprime.c (hunk 4 above)
Comment (fbaumdicker) [reply lib/msprime.c]: changed this throughout the code

File: lib/msprime.c (hunk 5 above)
Comment (fbaumdicker) [reply lib/msprime.c]: changed

File: lib/msprime.c (hunk 6 above)
Comment (fbaumdicker) [reply lib/msprime.c]: corrected everywhere

File: lib/msprime.c (hunk 7 above)
Comment (fbaumdicker) [reply lib/msprime.c]: corrected

File: msprime/cli.py (hunk 8 above)
Comment (fbaumdicker) [reply msprime/cli.py:196]: I see. But this is the part where the bug for the mspms output with breakpoints appeared. So going back to the original state would reintroduce this bug. I tried to be identical to the new version where this bug has been corrected.

Comment (jeromekelleher) [reply msprime/cli.py:196]: Ah, I see, sorry I got confused about where this PR was going. looks like I need to merge this and then rebase my PR against upstream.

**Comment by jeromekelleher** [comment]
Thanks @fbaumdicker, merging. I'll try and get the upstream PR updated and merged soon.

File: msprime/cli.py (hunk 8 above)
Comment (fbaumdicker) [reply msprime/cli.py]: Ok, so I will rebase this pull request stuff first and then we should be ready to merge this, correct?

**Comment by fbaumdicker** [comment]
Awesome! I am currently working on the statistical test so stay tuned for the next pull request.

**Pull request merged**